### PR TITLE
Run statistics

### PR DESCRIPTION
**Pull request by HenrikPoulsen**
I am making this a draft PR for now, since I expect this may not be the approach preferred.
So I have not made tests etc yet, I wanted to get some early feedback first so I can get a feel for what the appropriate way to solve this would be.

## Changes:

Collecting and logging run statistics. Such as what dependencies are out of date, what the state of the branch/PR is for each, the created/updated/closed time of the PR, things of that sort.
Also added an extra command line argument for dumping the stats to a json file, so it can be collected easily after a run.

## Context:

This is one potential solution for #2408

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/stats-reporter/index.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
+import { pkg } from '../../expose.cjs';
+import { logger } from '../../logger';
+import { BunyanRecord } from '../../logger/types';
```
Comment (viceice) [inline lib/util/stats-reporter/index.ts:3]:
```suggestion
import type { BunyanRecord } from '../../logger/types';
```

File: lib/util/stats-reporter/index.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+import * as memCache from '../cache/memory';
+import { readFile, writeFile } from '../fs';
+import { RenovateStats, RepositoryStats } from './types';
```
Comment (viceice) [inline lib/util/stats-reporter/index.ts:6]:
```suggestion
import type { RenovateStats, RepositoryStats } from './types';
```

File: lib/util/stats-reporter/types.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import { BunyanRecord } from '../../logger/types';
```
Comment (viceice) [inline lib/util/stats-reporter/types.ts:1]:
```suggestion
import type { BunyanRecord } from '../../logger/types';
```

File: lib/workers/branch/handle-existing.ts [hunk 4]
```diff
@@ -3,8 +3,11 @@ import { logger } from '../../logger';
 import { Pr, platform } from '../../platform';
 import { PrState } from '../../types';
 import { branchExists, deleteBranch } from '../../util/git';
+import { RepositoryStatisticsReporter } from '../../util/stats-reporter';
 import * as template from '../../util/template';
 import { BranchConfig } from '../types';
+//import { StatsReporter } from '../../util/stats-reporter';
+//import { DependencyStats } from '../../util/stats-reporter/types';
```
Comment (viceice) [inline lib/workers/branch/handle-existing.ts:10]: ?

**Review by viceice — COMMENTED** [review]
Looks promissing in general, but there a re alot open ends. config option ot yet used, new grapghql props not used ...

@rarkins WDYT?

File: lib/workers/branch/handle-existing.ts (hunk 4 above)
Comment (HenrikPoulsen) [reply lib/workers/branch/handle-existing.ts]: There has been some iterations on the code and I forgot to clean that up. It's gone now though. Thanks

**Comment by rarkins** [comment]
We have now adopted quite a few of these concepts into the main branch. Primarily:
- We have increased branch state in our cache
- We have a log message about branch state towards the end

I would therefore close this PR and concentrate our efforts on those.

**Pull request closed without merging**